bundles/org.openhab.binding.hive/src/main/java/org/openhab/binding/hive/internal/HiveHandlerFactory.java

```diff
@@ -0,0 +1,237 @@
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory;
+import org.eclipse.smarthome.io.net.http.HttpClientFactory;
+import org.openhab.binding.hive.internal.client.DefaultHiveClientFactory;
+import org.openhab.binding.hive.internal.client.HiveClientFactory;
+import org.openhab.binding.hive.internal.discovery.DefaultHiveDiscoveryService;
+import org.openhab.binding.hive.internal.handler.DefaultHiveAccountHandler;
+import org.openhab.binding.hive.internal.handler.DefaultHiveThingHandler;
+import org.openhab.binding.hive.internal.handler.strategy.*;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link HiveHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Ross Brown - Initial contribution
+ */
+@NonNullByDefault
+@Component(configurationPid = "binding.hive", service = ThingHandlerFactory.class)
+public final class HiveHandlerFactory extends BaseThingHandlerFactory {
+    private final Logger logger = LoggerFactory.getLogger(HiveHandlerFactory.class);
+
+    private final Set<ThingHandlerStrategy> strategiesForBoilerModule;
+    private final Set<ThingHandlerStrategy> strategiesForHeating;
+    private final Set<ThingHandlerStrategy> strategiesForHotWater;
+    private final Set<ThingHandlerStrategy> strategiesForHub;
+    private final Set<ThingHandlerStrategy> strategiesForThermostat;
+    private final Set<ThingHandlerStrategy> strategiesForTrvGroup;
+    private final Set<ThingHandlerStrategy> strategiesForTrv;
+
+    private final Map<ThingUID, ServiceRegistration<?>> discoveryServices = new HashMap<>();
+
+    private final HttpClient httpClient;
+    private final HiveClientFactory hiveClientFactory;
+
+    @Activate
+    public HiveHandlerFactory(@Reference final HttpClientFactory httpClientFactory) {
+        final HttpClient httpClient = httpClientFactory.createHttpClient("HiveBinding");
+        try {
+            httpClient.start();
+        } catch (Exception ex) {
+            // Wrap up all exceptions and throw a runtime exception as there
+            // is nothing we can do to recover.
+            throw new IllegalStateException("Could not start HttpClient.", ex);
+        }
```

Comment:
You should start this in the activate method instead.

Reply:
What is the reasoning for this? From my limited knowledge of OSGi I thought activate was always called immediately after the constructor.

I've reverted my change for now as I realised with my current implementation I need the `HttpClient` to be started before I can create an instance of `DefaultHiveClientFactory`.

If I move it I will need to either make `hiveClientFactory` non-final or put some lazy initialisation logic somewhere to make things work.

Reply:
My reasoning is that `activate` and `deactivate` methods should be used to handle service lifecycle. If for whatever reason osgi decides to reuse your service instance and reactivate with different configs, it should be able to. Not that I am an expert on how OSGI works (I'm more of a spring person), but it seems a better practice to make `activate` handle starting stuff, and `deactivate` handle stopping stuff since it seems like that was the design intent.

Reply:
As I said I'm quite new to OSGi but this is my understanding of how it works:

In an OSGi service's references can either "static" (default) or "dynamic" (opt-in).

Static references are only set at activation time. If a static reference _has_ to change the service is deactivated and a new one is created with the new configuration.

Dynamic references are at activation time but can be updated as the list of available implementations change by calling a `@Modified` method / `@Reference` setters.

The `@Activate` annotated constructor is an alternative to the `activate(...)` / `@Activate` method which only allows static references to be passed:

1. Component Activation Start
2. Call `@Activate` constructor
3. Call `@Reference` setters / Directly inject `@Reference` fields.
4. Call `@Activate` method
5. Component Activation End

Therefore I think that putting this code in the `@Activate` constructor is semantically the same as putting it in the `activate(...)` method. Any change to the static `HttpClientFactory` reference (and therefore `this.httpClient`) should trigger the service being deactivated and a replacement created.